### PR TITLE
Add Imitation Motor Controller 

### PR DESCRIPTION
**Pull request by amit-kumarh**
Firmware session 3 assignment

### CODE REVIEW

File: tutorials/fmc/fmc.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+volatile bool run_50hz = false;
+
+void timer0_callback(void) {
+    run_50hz = true;
+}
+
+void set_duty_cycle(int torque_request) {
+    float duty_cycle = (float)torque_request / 255.0;
+    OCR1B = (int16_t)(duty_cycle * 80.0);
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.c:18]: Magic numbers

File: tutorials/fmc/BUILD [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    srcs = [
+        "fmc.c",
+        "fmc.h",
+    ],
+    visibility = ["//visibility:public"],
+    deps = [
+        ":can_api",
+        "//libs/adc",
+        "//libs/timer",
+    ],
```
Comment (jack-greenberg) [inline tutorials/fmc/BUILD:15]: You might need to add `target_compatible_with`. Check other BUILD files for firmware for examples.

File: tutorials/fmc/fmc.c [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+int main(void) {
+    can_init(BAUD_500KBPS);
+    timer_init(&timer0_cfg);
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.c:23]: Where are you initializing timer1?

File: tutorials/fmc/fmc.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            // Over 2.5V for over one second but less than two
+            if (implausibility && implausibility_timer < 100) {
+                set_duty_cycle(0);
+                implausibility_timer++;
+                continue;
+            }
+
+            if (fmc.adc_value > 512) {
+                implausibility_timer++;
+                if (implausibility_timer >= 50) {
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.c:44]: Magic numbers

File: tutorials/fmc/fmc.c [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                    // Over 2.5V for more than 1 second
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.c:45]: Nice use of comments!

File: tutorials/fmc/fmc.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
+#include "libs/adc/api.h"
+#include "libs/timer/api.h"
+
+adc_pin_e adc_pin = ADC4;
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.h:4]: Maybe find a better variable name? `TEMPERATURE_SENSOR`?

File: tutorials/fmc/fmc.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+void timer0_callback(void);
+
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.h:7]: I recommend leaving comments about what these timers are doing and how you calculated the output compare match values and what they do. For both timers.

File: tutorials/fmc/fmc.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+name: fmc
+subscribe:
+  - name: motor_command
+    mob: 1
+publish:
+  - name: fmc
+    id: 0xC
+    freq_hz: 100
+    signals:
+      - name: adc_value
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.yml:10]: Maybe find a more descriptive name?

File: tutorials/fmc/fmc.yml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+        slice: 0 + 16
+        unit:
+          type: int16_t
+          name: ''
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.yml:14]: If it's temperature, maybe `degC` or just `C` or something.

File: tutorials/fmc/fmc.c [hunk 10]
```diff
@@ -0,0 +1,64 @@
+#include "fmc.h"
+#include "libs/adc/api.h"
+#include "libs/timer/api.h"
+#include "tutorials/fmc/can_api.h"
+
+#include <avr/interrupt.h>
+#include <avr/io.h>
+#include <stdio.h>
```
Comment (jack-greenberg) [inline tutorials/fmc/fmc.c:8]: Is this import needed? You might need `<stdbool.h>` for `bool`...

File: tutorials/fmc/BUILD [hunk 11]
```diff
@@ -0,0 +1,27 @@
+load("//bazel:defs.bzl", "cc_firmware")
+load("//bazel/tools:defs.bzl", "can_api_files")
+
+cc_firmware(
+    name = "fmc",
+    srcs = [
+        "fmc.c",
+        "fmc.h",
```
Comment (jack-greenberg) [inline tutorials/fmc/BUILD:8]: Maybe `fmc_config.h`?

**Review by jack-greenberg — CHANGES_REQUESTED** [review]

File: tutorials/fmc/fmc.c (hunk 10 above)
Comment (amit-kumarh) [reply tutorials/fmc/fmc.c:8]: stdio would be for the int16_t right

File: tutorials/fmc/fmc.c [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    can_receive_motor_command();
+    while (1) {
+        fmc.temperature = adc_read(TEMPERATURE_SENSOR);
+
+        if (run_50hz) {
+            // Over 2.5V for over one second but less than two
+            if (implausibility && implausibility_timer < TWO_SECONDS) {
+                set_duty_cycle(0);
+                implausibility_timer++;
+                continue;
```
Comment (amit-kumarh) [inline tutorials/fmc/fmc.c:41]: whoops, this is bad. run_50Hz doesn't get set to false! and we don't send the CAN message

Comment (amit-kumarh) [reply tutorials/fmc/fmc.c:41]: same on line 49

**Pull request closed without merging**